### PR TITLE
failureMap for plugins

### PR DESCRIPTION
**Pull request by novaknole**
## Description

- Adds failureMap feature to plugins.

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/contracts/voting/admin/Admin.sol [hunk 1]
```diff
@@ -44,9 +44,11 @@ contract Admin is PluginCloneable, Proposal {
     /// @notice Creates and executes a new proposal.
     /// @param _metadata The metadata of the proposal.
     /// @param _actions The actions to be executed.
+    /// @param _allowFailureMap Allows proposal to succeed even if an action reverts. Uses bitmap representation. If the bit at index `x` is 1, the tx succeeds even if the action at `x` failed. Passing 0 will be treated as atomic execution.
     function executeProposal(
         bytes calldata _metadata,
-        IDAO.Action[] calldata _actions
+        IDAO.Action[] calldata _actions,
+        uint256 _allowFailureMap
```
Comment (brickpop) [inline packages/contracts/contracts/voting/admin/Admin.sol:51]: Does it make sense that the admin plugin needs this? If execution will be immediate, it might be cheaper to save one parameter and have the admin see that execution will fail before submitting it. The admin can split actions into different proposals with low overhead, unlike any governance plugin.

Happy to hear you view on this, but feel free to resolve if you recommend adding this here.

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 2]
```diff
@@ -168,15 +168,15 @@ describe('AddresslistVoting', function () {
       await expect(
         voting
           .connect(signers[1])
-          .createProposal(dummyMetadata, [], 0, 0, VoteOption.None, false)
+          .createProposal(dummyMetadata, [], 0, 0, 0, VoteOption.None, false)
```
Comment (brickpop) [inline packages/contracts/test/voting/addresslist-voting.ts:171]: No new tests to check how the failure bitmap works? Separate task?

File: packages/subgraph/tests/constants.ts [hunk 3]
```diff
@@ -36,6 +36,10 @@ export const START_DATE = '1644851000';
 export const END_DATE = '1644852000';
 export const SNAPSHOT_BLOCK = '100';
 
+// Use 1 for testing as default value is anyways 0 
+// and test might succeed even though it shouldn't
+export const ALLOW_FAILURE_MAP = '1';
```
Comment (brickpop) [inline packages/subgraph/tests/constants.ts:41]: `1` means that the first action can fail, is that correct? Won't this change the current test behavior?

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 2 above)
Comment (novaknole) [reply packages/contracts/test/voting/addresslist-voting.ts:171]: https://github.com/aragon/core/pull/207/files#diff-2a15c769379f924880f7d57bdc744281b5db84e22fcf521247a06747a7e5966fR285 Check this. The same thing is in every plugin.

File: packages/subgraph/tests/constants.ts (hunk 3 above)
Comment (novaknole) [reply packages/subgraph/tests/constants.ts:41]: I only pass `1` in a test where actions are []`. We anyways don't test voting plugins as integration tests(also testing dao.execute results). So I only use `1` just to test the fact that it was stored correctly in a voting plugin and was emitted correctly.

File: packages/contracts/contracts/voting/admin/Admin.sol (hunk 1 above)
Comment (novaknole) [reply packages/contracts/contracts/voting/admin/Admin.sol:51]: It's still needed. Imagine, admin wants to execute, but is not sure whether or not 2nd action fails or not, but he doesn't want the overal tx to fail.  So uses bitmap param.

Really doesn't increase gas that much.

File: packages/subgraph/tests/token-voting/utils.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import {Address, BigInt, Bytes, ethereum} from '@graphprotocol/graph-ts';
+import {Address, BigInt, Bytes, ethereum, log} from '@graphprotocol/graph-ts';
```
Comment (mathewmeconry) [inline packages/subgraph/tests/token-voting/utils.ts:1]: remove `log`

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

**Review by mathewmeconry — DISMISSED** [review]

File: packages/contracts/contracts/voting/admin/Admin.sol (hunk 1 above)
Comment (brickpop) [reply packages/contracts/contracts/voting/admin/Admin.sol:50]: The main goal of the admin plugin is to gain speed and save gas, compared to the other ones.

If the gas savings are light, then feel free to resolve the issue, but otherwise I would recommend to save one parameter and make execution atomic as it was.

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 2 above)
Comment (brickpop) [reply packages/contracts/test/voting/addresslist-voting.ts:171]: Right, but I mean something like `0b0000101` and then, checking that actions 1 and 3 can fail without execution actually reverting.

Same with actions 2, 4, 5, 6, 7, 8 failing, provoking the whole execution to fail.

We can create a separate task to add this later, but it has to exist

Comment (novaknole) [reply packages/contracts/test/voting/addresslist-voting.ts:171]: I don't think thats the test that should be here . I already wrote them in `dao.ts` ! for these tests, you only care about if event was emitted with a correct failureMap.  Otherwise, you really are testing this with kind of e2e or integration.

feel free to create a task if you still want.

**Review by Michael-A-Heuer — COMMENTED** [review]
A couple of comments, mostly NatSpec and input argument ordering

File: packages/contracts/contracts/voting/admin/Admin.sol [hunk 5]
```diff
@@ -43,9 +43,11 @@ contract Admin is PluginCloneable, Proposal {
...
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/admin/Admin.sol:50]: In ProposalBase, we do
```solidity
    function _createProposal(
        address _creator,
        bytes calldata _metadata,
        uint64 _startDate,
        uint64 _endDate,
        IDAO.Action[] calldata _actions,
        uint256 _allowFailureMap
    )
```

I would change in MajorityVotingBase
```solidity
function createProposal(
        bytes calldata _metadata,
        IDAO.Action[] calldata _actions,
        uint256 _allowFailureMap,
        uint64 _startDate,
        uint64 _endDate,
        VoteOption _voteOption,
        bool _tryEarlyExecution
    )
```
to
```solidity
function createProposal(
        bytes calldata _metadata,
        uint64 _startDate,
        uint64 _endDate,
        IDAO.Action[] calldata _actions,
        uint256 _allowFailureMap,
        VoteOption _voteOption,
        bool _tryEarlyExecution
    )
```

[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/admin/Admin.sol:50]: Maybe we can create a follow-up task @brickpop

[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/admin/Admin.sol:50]: Asked for a follow-up task here https://discord.com/channels/672466989217873929/950422827893420153/1067387013239668747
to resolve this for this PR

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 6]
```diff
@@ -178,21 +182,24 @@ contract Multisig is PluginUUPSUpgradeable, ProposalUpgradeable, Addresslist {
...
     /// @param _metadata The metadata of the proposal.
     /// @param _actions The actions that will be executed after the proposal passes.
+    /// @param _allowFailureMap Allows proposal to succeed even if an action reverts. Uses bitmap representation. If the bit at index `x` is 1, the tx succeeds even if the action at `x` failed. Passing 0 will be treated as atomic execution.
     /// @param _approveProposal If `true`, the sender will approve the proposal.
     /// @param _tryExecution If `true`, execution is tried after the vote cast. The call does not revert if early execution is not possible.
     /// @return proposalId The ID of the proposal.
     function createProposal(
         bytes calldata _metadata,
         IDAO.Action[] calldata _actions,
+        uint256 _allowFailureMap,
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/multisig/Multisig.sol:202]: Asked for a follow-up task here https://discord.com/channels/672466989217873929/950422827893420153/1067387013239668747
to resolve this for this PR

File: packages/contracts/contracts/voting/addresslist/AddresslistVoting.sol [hunk 7]
```diff
@@ -76,6 +76,7 @@ contract AddresslistVoting is Addresslist, MajorityVotingBase {
     function createProposal(
         bytes calldata _metadata,
         IDAO.Action[] calldata _actions,
+        uint256 _allowFailureMap,
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/addresslist/AddresslistVoting.sol:79]: Asked for a follow-up task here https://discord.com/channels/672466989217873929/950422827893420153/1067387013239668747
to resolve this for this PR

File: packages/contracts/contracts/voting/token/TokenVoting.sol [hunk 8]
```diff
@@ -60,6 +60,7 @@ contract TokenVoting is MajorityVotingBase {
     function createProposal(
         bytes calldata _metadata,
         IDAO.Action[] calldata _actions,
+        uint256 _allowFailureMap,
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/token/TokenVoting.sol:63]: Asked for a follow-up task here https://discord.com/channels/672466989217873929/950422827893420153/1067387013239668747
to resolve this for this PR

**Review by Michael-A-Heuer — APPROVED** [review]

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 9]
```diff
@@ -305,13 +305,13 @@ contract Multisig is PluginUUPSUpgradeable, ProposalUpgradeable, Addresslist {
...
     /// @return actions The actions to be executed in the associated DAO after the proposal has passed.
-    /// @return allowFailureMap The bit map representations of which actions are allowed to revert so tx still succeeds.
+    /// @param allowFailureMap A bitmap allowing the proposal to succeed, even if individual actions might revert. If the bit at index `i` is 1, the proposal succeeds even if the `i`th action reverts. A failure map value of 0 requires every action to not revert.
     function getProposal(uint256 _proposalId)
         public
         view
         returns (
             bool executed,
-            uint32 approvals,
+            uint16 approvals,
```
Comment (brickpop) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:314]: Why this change?

**Review by brickpop — APPROVED** [review]
